feat: step3 레이싱카 게임 구현

ddoriya-dev opened the pull request.

Commit:
feat: step3 레이싱카 게임 구현

ddaaac commented on src/main/kotlin/study/racing/RacingCarGame.kt:
```diff
@@ -0,0 +1,53 @@
+    private val gameSettings: GameSettings,
+) {
+    private val random = Random
+    private var carList = mutableListOf<Car>()
```
[`var와 val` 그리고 `List와 MutableList`](https://velog.io/@hyemdooly/Kotlin%EC%9D%98-var-List-vs-val-MutableList)는 어떻게 다를까요?

ddoriya-dev replied:
가변객체에 굳이 불필요하게 var를 사용하였군요 ㅠㅠ 
감사합니다.

ddaaac commented on src/main/kotlin/study/racing/RacingCarGame.kt:
```diff
@@ -0,0 +1,53 @@
+class RacingCarGame(
+    private val gameSettings: GameSettings,
+) {
+    private val random = Random
```
kotlin Random의 구현체는 싱글톤 방식으로 제공되어서 따로 변수로 뺄 필요없이 바로 Random.nextInt() 등을 활용해도 좋습니다.
만약 시드값을 직접 줘야할 필요가 있다면 그 때 변수로 빼도 좋을 것 같네요.

ddoriya-dev replied:
아 그렇군요. 시드값을 주는 방식으로 매번하다보니 굳이 인스턴스를 생성할 필요가 없겠군요~ 감사합니다.

ddaaac commented on src/main/kotlin/study/racing/RacingCarGame.kt:
```diff
@@ -0,0 +1,53 @@
+
+    private fun init() {
+        this.carList = mutableListOf()
+        repeat(gameSettings.carCount) {
```
repeat 활용해주셨군요 👍

ddaaac commented on src/main/kotlin/study/racing/RacingCarGame.kt:
```diff
@@ -0,0 +1,53 @@
+
+    private fun playGame() {
+        carList.forEach {
+            it.position += move()
```
자동차를 움직이는 방법에는 다양한 방법이 있을 것 같아요.
car.move()
car.position += xx
car.move(xx)

등등 의 여러 방법 중에서 상준님이 이 방법을 사용해주신 이유가 궁금합니다!

ddoriya-dev replied:
네네 move() 라는걸 Car 클래스에 넣는부분으로 개선하였습니다~

생각해보니 밖으로 빼는부분은 안좋은 부분이군요..

ddaaac replied:
어떤 방법이 안 좋다라는 의미의 코멘트는 아니었습니다!

상준님이 어떤 의도를 갖고 코드를 구현했는지 궁금해서 남긴 코멘트였어요 😄

ddaaac commented on src/main/kotlin/study/racing/RacingCarGame.kt:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    private fun move(): Int {
+        return if (random.nextInt(GameRule.MIN_RANDOM_POSITION, GameRule.MAX_RANDOM_POSITION) > 3) 1 else 0
```
상수를 다른 클래스에서 보관하는 것과 이 클래스에서 관리하는 것은 어떤 차이가 있을까요?

ddoriya-dev replied:
아무래도 Car 클래스에 보관하는것이 더좋아보이네요. 감사합니다.

ddaaac replied:
이 코멘트가 반영이 되지 않아보여요. 확인 부탁드립니다~

ddoriya-dev replied:
해당 move() 메서드를 Car 클래스로 이동하였습니다~

ddaaac commented on src/main/kotlin/study/racing/RacingCarGameSettings.kt:
```diff
@@ -0,0 +1,21 @@
+
+data class GameSettings(
+    var carCount: Int,
+    var racingCount: Int,
```
가변변수로 만들어주신 이유가 있을까요?

ddoriya-dev replied:
사실 Settings 를 의미하기때문에 다시 설정이 변경될 가능성을 가져서 가변변수로 진행하였습니다.
하지만 해당 부분에는 큰의미가없어 보이긴하네요..

ddaaac commented on src/test/kotlin/study/racing/RacingCarGameTest.kt:
```diff
@@ -0,0 +1,51 @@
+/**
+ * @author 이상준
+ */
+class RacingCarGameTest : StringSpec({
```
UI를 담당하는 클래스에 대해서는 테스트를 하지 않아도 된다는 요구사항이 있습니다.
핵심 로직에 대한 테스트를 작성해봐도 좋을 것 같아요!

ddoriya-dev replied:
네네! 해당부분은 제외하고 핵심로직에 구현핤효ㅜ있도록 진행하겠습니다.

ddaaac commented on src/main/kotlin/study/racing/Car.kt:
```diff
@@ -0,0 +1,6 @@
+/**
+ * @author 이상준
+ */
+data class Car(var position: Int = 1)
```
외부에서 객체 내부의 값을 변경할 수 있다면 어떤 일들이 생길 수 있을까요?

이를 해결하기 위해서는 어떤 방법들 [1](https://kotlinlang.org/docs/properties.html#getters-and-setters) [2](https://discuss.kotlinlang.org/t/private-setter-for-var-in-primary-constructor/3640/9)이 있을까요?

ddoriya-dev replied:
아 감사합니다 해당부분을 놓쳤군요 ㅠㅠ

ddaaac reviewed (changes requested):
안녕하세요 상준님!

구현 깔끔하게 잘 해주셨네요.

몇 가지 코멘트 남겼으니 확인해주시고 궁금하신 부분이 있다면 언제든지 dm 주세요.

ddoriya-dev reviewed (commented):
놓치고있는 부분이 많았군요 ㅠㅠ
리뷰 감사합니다~ 코드 수정하여 다시 전달드립니다~

Commit:
refactor: 리뷰에 따른 코드 개선

https://github.com/next-step/kotlin-racingcar/pull/1645#pullrequestreview-2441290157

ddaaac commented on src/main/kotlin/study/racing/Car.kt:
```diff
@@ -0,0 +1,20 @@
+ * @author 이상준
+ */
+class Car {
+    private var position: Int = 0
+
+    fun getPosition(): Int {
+        return this.position
+    }
```
[private setter](https://ksabs.tistory.com/247)를 활용해봐도 좋을 것 같아요!

ddoriya-dev replied:
네 감사합니다! 그러한 방법이있군요 ㅠㅠ

ddaaac commented on src/main/kotlin/study/racing/RacingCarGame.kt:
```diff
@@ -0,0 +1,45 @@
+class RacingCarGame(
+    private val gameSettings: GameSettings,
+) {
+    private val carList = mutableListOf<Car>()
+
+    private fun init() {
+        repeat(gameSettings.carCount) {
+            this.carList.add(Car())
```
kotlin에서 [리스트 생성](https://www.baeldung.com/kotlin/initialize-list)엔 여러 방법이 있습니다.
잘 활용해본다면 가변 리스트를 사용하지 않아도 될 것 같아요.

ddoriya-dev replied:
가변리스트 외 어떠한 방법있을지.. 힌트를 조금더 주실수있으실까요 ㅠㅠ

ddaaac commented on src/main/kotlin/study/racing/Car.kt:
```diff
@@ -0,0 +1,20 @@
+        return this.position
+    }
+
+    fun move() {
+        if (Random.nextInt(GameRule.MIN_RANDOM_POSITION, GameRule.MAX_RANDOM_POSITION) > 3) {
+            this.position += 1
```
다음 단계를 진행하면서 이 부분을 어떻게 테스트할지 고민해보셔도 좋을 것 같습니다~

ddoriya-dev replied:
넵 다음단계에서는 조금더 개선해보록 하겠습니다~

ddaaac reviewed (approved):
안녕하세요 상준님 👋 

리뷰 반영 깔끔하게 잘 해주셨네요.

몇 가지 코멘트 확인해주시고 다음 단계 진행해주세요.
